static_frame/core/axis_map.py

```diff
@@ -33,6 +35,46 @@ def get_extractor(
     return lambda x: x
 
 
+def _bus_to_hierarchy_inner_hierarchies(
+        bus: tp.Union[Bus, 'Yarn'],
+        axis: int,
+        extractor: tp.Callable[[IndexBase], IndexBase],
+        init_exception_cls: tp.Type[Exception],
+        ) -> tp.Tuple[IndexHierarchy, IndexBase]:
+    '''
+    Specialized version of :func:`bus_to_hierarchy` for the case where Bus's frames contains only hierarchical indices on the axis of concatentation
+    '''
+    opposite: tp.Optional[IndexBase] = None
+
+    def get_next(items_iter: tp.Iterator[tp.Tuple[tp.Hashable, Frame]]) -> IndexHierarchy:
+        nonlocal opposite
+        label, frame = next(items_iter)
+
+        if axis == 0:
+            axis0, axis1 = extractor(frame.index), frame.columns
+        else:
+            assert axis == 1
+            axis0, axis1 = extractor(frame.columns), frame.index
+
+        if opposite is None:
+            opposite = extractor(axis1)
+        else:
+            if not opposite.equals(axis1):
+                raise init_exception_cls('opposite axis must have equivalent indices')
+
+        assert isinstance(axis0, IndexHierarchy) # true assert
+        return axis0.level_add(label)
+
+    items_iter = iter(bus.items())
+
+    primary = IndexHierarchyGO(get_next(items_iter))
+
+    for _ in range(1, len(bus)):
+        primary.extend(get_next(items_iter))
+
+    return IndexHierarchy(primary), opposite # type: ignore
```

Comment:
I think it is better if the function processes pairs only, and does not do the iteration. Then we get a more natural iteration at the end. Also, I think we can remove some asserts and replace plausible errors with exceptions. Finally, since this is an internally managed IHg, we can strip out its TypeBlocks for usage in our final returned IH.

```suggestion
    def level_add(tp.Tuple[label: tp.Hashable, frame=Frame]) -> IndexHierarchy:
        nonlocal opposite

        if axis == 0:
            axis0, axis1 = extractor(frame.index), frame.columns
        else:
            axis0, axis1 = extractor(frame.columns), frame.index

        if opposite is None:
            opposite = extractor(axis1)
        else:
            if not opposite.equals(axis1):
                raise init_exception_cls('opposite axis must have equivalent indices')

        if axis0._NDIM != 2:
             raise RuntimeError('bad index') # something better!
        return axis0.level_add(label)

    items_iter = iter(bus.items())

    primary = IndexHierarchyGO(level_add(next(items_iter)))

    for level in items_iter:
        primary.extend(level_add(level))

    return IndexHierarchy._from_type_blocks(primary._blocks, ), opposite # type: ignore
```

Reply:
Good suggestion! One small note: `IndexHierarchy(primary)` is much faster than `IndexHierarchy._from_type_blocks(primary._blocks)`. This is because you have to re-create a bunch of different components inside of `_from_type_blocks`, whereas the specialized path in `__init__` can borrow all references to static internally managed components without having to do any analysis